tests/system/test_client.py

```diff
@@ -102,3 +102,41 @@ def test_download_blob_to_file_w_uri(
             stored_contents = file_obj.read()
 
     assert stored_contents == payload
+
+
+def test_download_blob_to_file_w_etag(
+    storage_client, shared_bucket, blobs_to_delete, service_account,
+):
+    filename = "kittens"
+    blob = shared_bucket.blob(filename)
+    payload = b"fluffy"
+    blob.upload_from_string(payload)
+    blobs_to_delete.append(blob)
+
+    with tempfile.NamedTemporaryFile() as temp_f:
+        with open(temp_f.name, "wb") as file_obj:
+            with pytest.raises(exceptions.NotModified):
+                storage_client.download_blob_to_file(
+                    "gs://" + shared_bucket.name + "/" + filename,
+                    file_obj,
+                    if_etag_not_match=blob.etag,
+                )
+
+            with pytest.raises(exceptions.PreconditionFailed):
+                storage_client.download_blob_to_file(
+                    "gs://" + shared_bucket.name + "/" + filename,
+                    file_obj,
+                    if_etag_match="kittens",
+                )
```

Comment:
Because they share the same `file_obj`, the two successfull cases actually end up concatenating the payload twice, which is odd.  I think it would be better to split the four cases apart, using separate buffers for each:  since we are passing a file object, and not a filename, we can create `io.BytesIO` instances.  For the two cases which succeed, we can then assert that `buffer.getvalue() matches the expected payload.

Reply:
Yes, I did realize that but didn't really care since I wasn't checking the contents (as there are already tests that cover that). Will change as you suggest though. I'm assuming "split the four cases apart" doesn't mean to actually break them into four separate tests, as there is overhead to create the blob and such. If you do want that just let me know.

Reply:
One testcase function is still fine here -- we can just create new buffers for each of the four invocations.